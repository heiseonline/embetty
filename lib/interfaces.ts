// eslint-disable-next-line @typescript-eslint/ban-types
export type Constructor<T = {}> = new (...args: any[]) => T
