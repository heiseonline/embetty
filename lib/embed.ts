
// @ts-ignore
import EMBETTY_LOGO from '!raw-loader!../assets/embetty.svg'

export default class Embed extends window.HTMLElement {
  _fetched = false
  url: any
  _data: any

  get serverUrl() {
    const baseUrl = document.querySelector<HTMLElement>(
      'meta[data-embetty-server]'
    )
    return (
      this.getAttribute('server-url') ||
      (baseUrl && baseUrl.dataset.embettyServer) ||
      ''
    )
  }

  isVisible() {
    return window.getComputedStyle(this).display !== 'none'
  }

  _api(url: string) {
    return this.serverUrl + url
  }

  emit(name: string, data?: any) {
    this.dispatchEvent(new window.CustomEvent(name, data))
  }

  on(name: string, cb: EventListenerOrEventListenerObject) {
    this.addEventListener(name, cb, false)
  }

  async becomesVisible() {
    if (this.url) await this.fetchData()
    this.shadowRoot!.innerHTML = this.render()

    this.emit('initialized')
  }

  async connectedCallback() {
    this.attachShadow({ mode: 'open' })
  }

  get wrapper() {
    return this.shadowRoot!.querySelector('.wrapper')
  }

  get embettyLogo() {
    return EMBETTY_LOGO
  }

  async fetchData() {
    const _response = await window.fetch(this.url)
    this._data = await _response.json()
    this._fetched = true
  }

  render() {
    // @ts-ignore
    return this.constructor.compiledTemplate.render(this)
  }
}
