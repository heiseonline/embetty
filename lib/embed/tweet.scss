@import '../element';
@import 'powered-by';

$profile-image-width: 36px;
$quoteLineWidth: 4px;

:host(.answered) {
  margin-top: 0;
  margin-bottom: 0.5rem;
  border: 0;
  padding: 0;

  header {
    padding-bottom: 0.5rem;
  }

  article {
    border-left: $quoteLineWidth solid #bbb;
    margin-left: $profile-image-width / 2 - $quoteLineWidth / 2;
    padding-left: 2rem;
    padding-bottom: 1rem;

    p {
      font-size: 14px;
    }

    #created-at {
      display: none;
    }
  }

  #powered-by {
    display: none;
  }
}

:host {
  position: relative;
  overflow: hidden;
  max-width: 642px;
  padding: 1rem 1.2rem;

  @media (min-width: 600px) {
    padding: 1.5rem 2rem;
  }

  header {
    display: flex;
    align-items: center;
    margin-bottom: .5rem;

    img {
      width: $profile-image-width;
      height: $profile-image-width;
      border-radius: 50%;
    }

    > span {
      display: inline-block;
      margin: 0 var(--embetty-spacing, $embetty-spacing);
    }

    strong {
      font-size: 16px;
      display: block;
    }

    a,
    a:hover {
      font-size: 14px;
      color: #697882;
      text-decoration: none;
    }
  }

  article {
    span {
      display: block;
    }

    p {
      margin: 0 auto 0.5rem;
      line-height: 1.4;
      letter-spacing: .01em;

      @media (min-width: 600px) {
        font-size: 18px;
      }

      a {
        color: #012469;
        text-decoration: none;
      }

      a:hover {
        text-decoration: underline;
      }
    }
  }

  #media {
    a {
      display: block;
      height: 100%;
      display: flex;
      align-items: center;

      &:not(:first-child) {
        display: none;
      }
    }

    img {
      max-width: 100%;
      width: 100%;
      height: 100%;
      object-fit: cover;
    }

    @media (min-width: 600px) {
      display: grid;
      grid-column-gap: 1px;
      grid-row-gap: 1px;

      a:not(:first-child) {
        display: block;
      }

      &.media-2 {
        grid-template-columns: 50% 50%;
      }

      &.media-3 {
        grid-template-columns: auto 40%;

        a:first-child {
          grid-row: 1 / span 2;
        }
      }

      &.media-4 {
        grid-template-columns: auto 20%;

        a:first-child {
          grid-row: 1 / span 3;
        }
      }
    }
  }

  #links {
    border: 1px solid var(--embetty-border-color, $embetty-border-color);
    border-width: 1px;
    border-radius: 4px;
    text-decoration: none;
    display: flex;
    flex-direction: column;
    color: #14171a;
    font-size: 14px;

    &:hover {
      background-color: rgb(245, 248, 250);
      border-color: rgba(136,153,166,.5);
      transition: background-color .15s ease-in-out, border-color .15s ease-in-out;
    }

    @media (min-width: 600px) {
      flex-direction: row;
    }

    img {
      max-width: 100%;
      object-fit: cover;
      display: inline-block;

      @media (min-width: 600px) {
        height: 125px;
        width: 125px;
        min-width: 125px;
      }
    }

    & > *:last-child {
      margin-bottom: 0;
    }

    #link-body {
      padding: .5rem;

      @media (min-width: 600px) {
        display: flex;
        flex-direction: column;
        padding: .5rem .8rem;
      }
    }

    h3 {
      font-size: 14px;
      line-height: 1.3;
      margin: 0;
      margin-bottom: .3em;
    }

    p {
      display: none;

      @media (min-width: 600px) {
        display: block;
        flex-grow: 1;
        hyphens: auto;
        line-height: 18px;
        font-size: 14px;
        margin: 0;
        margin-bottom: .3em;
      }
    }

    span {
      margin-top: auto;
      color: #999;
    }
  }

  #created-at {
    margin-top: .5rem;
    display: block;
    font-size: 14px;
    color: #777;
    text-decoration: none;

    svg {
      height: 22px;
      vertical-align: middle;
    }
  }
}
