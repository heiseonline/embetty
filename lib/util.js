import hogan from 'hogan.js'

export function wait(ms = undefined) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

export function computedStyle(element, key) {
  return window.getComputedStyle(element)[key]
}

export function marginHeight(element) {
  const top = parseFloat(computedStyle(element, 'margin-top'))
  const bottom = parseFloat(computedStyle(element, 'margin-bottom'))
  return top + bottom
}

export function height(element) {
  return parseFloat(computedStyle(element, 'height')) + marginHeight(element)
}

export function parseHostname(url) {
  const a = document.createElement('a')
  a.setAttribute('href', url)
  return a.hostname
}

export function createTemplate(content, css) {
  const html = `<style>${css}</style>${content}`
  const template = document.createElement('template')
  template.innerHTML = html
  return template
}

export function defineElement(name, Class) {
  const template = createTemplate(Class.template, Class.css)
  window.ShadyCSS.prepareTemplate(template, name)

  Class.compiledTemplate =
    Class.compiledTemplate || hogan.compile(Class.template)

  if (window.WebComponents?.ready) {
    window.customElements.define(name, Class)
  } else {
    window.addEventListener('WebComponentsReady', function() {
      window.customElements.define(name, Class)
    })
  }
}
