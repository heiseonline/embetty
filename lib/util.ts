import hogan from 'hogan.js'

export function wait(ms = undefined) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export function computedStyle(
  element: Element,
  key: keyof CSSStyleDeclaration
) {
  return window.getComputedStyle(element)[key]
}

export function marginHeight(element: Element) {
  const top = Number.parseFloat(computedStyle(element, 'marginTop') as string)
  const bottom = Number.parseFloat(
    computedStyle(element, 'marginBottom') as string
  )
  return top + bottom
}

export function height(element: Element) {
  return (
    Number.parseFloat(computedStyle(element, 'height') as string) +
    marginHeight(element)
  )
}

export function parseHostname(url: string) {
  const a = document.createElement('a')
  a.setAttribute('href', url)
  return a.hostname
}

export function createTemplate(content: string, css: string) {
  const html = `<style>${css}</style>${content}`
  const template = document.createElement('template')
  template.innerHTML = html
  return template
}

// eslint-disable-next-line sonarjs/cognitive-complexity
export function defineElement(name: string, Class: any) {
  const template = createTemplate(Class.template, Class.css)
  window.ShadyCSS.prepareTemplate(template, name)

  Class.compiledTemplate =
    Class.compiledTemplate || hogan.compile(Class.template)
  // @ts-ignore
  if (window.WebComponents.ready) {
    if (!window.customElements.get(name)) {
      window.customElements.define(name, Class)
    }
  } else {
    window.addEventListener('WebComponentsReady', function () {
      if (!window.customElements.get(name)) {
        window.customElements.define(name, Class)
      }
    })
  }
}
