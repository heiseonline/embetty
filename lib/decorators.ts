import { defineElement } from './util'

export function webcomponent(tag: string, options?: ElementDefinitionOptions) {
  return (target: any) => {
    defineElement(tag, target)
  }
}
