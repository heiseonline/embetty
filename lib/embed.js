import { wait } from './util'
import urljoin from 'url-join'

// eslint-disable-next-line
const EMBETTY_LOGO = require('!raw-loader!../assets/embetty.svg').default.toString()

export default class Embed extends window.HTMLElement {
  constructor() {
    super()
    this._fetched = false
  }

  get serverUrl() {
    const baseUrl = document.querySelector('meta[data-embetty-server]')
    return this.getAttribute('server-url') ||
      (baseUrl && baseUrl.dataset.embettyServer) ||
      ''
  }

  isVisible() {
    return window.getComputedStyle(this).display !== 'none'
  }

  _api(url) {
    return urljoin(this.serverUrl, url)
  }

  emit(name, data) {
    this.dispatchEvent(new window.CustomEvent(name, data))
  }

  on(name, cb) {
    this.addEventListener(name, cb, false)
  }

  async becomesVisible() {
    if (this.url) await this.fetchData()
    this.shadowRoot.innerHTML = this.render()

    await wait() // wait for ShadyDOM to render
    window.ShadyCSS.styleElement(this)
    this.emit('initialized')
  }

  async connectedCallback() {
    this.attachShadow({ mode: 'open' })
    await wait()
  }

  get wrapper() {
    return this.shadowRoot.querySelector('.wrapper')
  }

  get embettyLogo() {
    return EMBETTY_LOGO
  }

  async fetchData() {
    const _response = await window.fetch(this.url)
    this._data = await _response.json()
    this._fetched = true
  }

  render() {
    return this.constructor.compiledTemplate.render(this)
  }
}
