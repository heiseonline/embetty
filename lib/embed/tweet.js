import { defineElement, height, parseHostname } from '../util'
import Embed from '../embed'
import Observable from '../observable'

const CSS = require('!css-loader!postcss-loader!sass-loader!./tweet.scss').toString()

const HEIGHT_OFFSET = 2

class Tweet extends Observable(Embed) {
  constructor(status, parent, options = {}) {
    super()
    if (status) this.setAttribute('status', status)
    if ('include-thread' in options) this.setAttribute('include-thread', '')
    this.parent = parent

    this.on('initialized', () => {
      this.appendAnsweredTweet()
    })
    this.on('initialized', () => {
      this.fitCardHeight()
    })
  }

  appendAnsweredTweet() {
    if (!this.isReply || !this.hasAttribute('include-thread')) return
    const answered = new Tweet(this.answeredTweetId, this, {
      'include-thread': '',
    })
    answered.classList.add('answered')
    this.shadowRoot.insertBefore(answered, this.shadowRoot.firstChild)
  }

  // TODO: Reduce complexity of fitCardHeight()
  // eslint-disable-next-line sonarjs/cognitive-complexity, complexity
  fitCardHeight() {
    if (!this.hasLinks) return

    const section = this.shadowRoot.querySelector('#links')
    const linkBody = this.shadowRoot.querySelector('#link-body')

    if (section.clientWidth === linkBody.clientWidth) return

    const p = section.querySelector('p')
    if (!p) return

    const imgHeight = height(section.querySelector('img'))
    let counter = 0
    let last = ''

    // eslint-disable-next-line no-loops/no-loops
    while (height(section) - HEIGHT_OFFSET > imgHeight) {
      if (++counter > 200) break
      if (last === p.textContent) break
      last = p.textContent
      p.textContent = `${p.textContent.replace(/\W*\s(\S)*$/, '')}â€¦`
    }
  }

  get answeredTweets() {
    return this.shadowRoot.querySelectorAll('embetty-tweet')
  }

  get url() {
    return this._api(`/tweet/${this.getAttribute('status')}`)
  }

  get twitterUrl() {
    return `https://twitter.com/${this._data.user.screen_name}/status/${this._data.id_str}`
  }

  get createdAt() {
    const createdAt = this._data.created_at.replace(/\+\d{4}\s/, '') // FU IE11
    return new Date(createdAt)
  }

  get profileImageUrl() {
    return `${this.url}-profile-image`
  }

  get linkImageUrl() {
    return `${this.url}-link-image`
  }

  get fullText() {
    return this._data.full_text
      .replace(/#([^\s-|]+)/g, (hashTag, word) => {
        return `<a href="https://twitter.com/hashtag/${word}">${hashTag}</a>`
      })
      .replace(/@(\w+)/g, (name, word) => {
        return `<a href="https://twitter.com/${word}">${name}</a>`
      })
      .replace(/(https:\/\/\S+)$/, link => {
        if (this.hasMedia && this.media[0].url === link) return ''
        return link
      })
  }

  get userName() {
    return this._data.user.name
  }

  get screenName() {
    return this._data.user.screen_name
  }

  get media() {
    const extended = this._data.extended_entities || {}
    const media = extended.media || []
    return media.map((m, idx) => {
      m.imageUrl = `${this.url}-images-${idx}`
      return m
    })
  }

  get hasMedia() {
    return this.media.length > 0
  }

  get links() {
    return this._data.entities.urls
  }

  get link() {
    return this.links[0]
  }

  get linkHostname() {
    return parseHostname(this.link.url)
  }

  get hasLinks() {
    return this.links.length > 0
  }

  get answeredTweetId() {
    return this._data.in_reply_to_status_id_str
  }

  get isReply() {
    return !!this.answeredTweetId
  }

  static get css() {
    return CSS
  }

  static get template() {
    return `
    <style>${CSS}</style>
      <header>
        <img src="{{profileImageUrl}}" />
        <span>
          <strong>{{userName}}</strong>
          <a href="https://twitter.com/{{screenName}}" target="_blank" rel="noopener" >@{{screenName}}</a>
        </span>
      </header>
      <article>
        <p>{{{ fullText }}}</p>
        {{#hasMedia}}
          <section class="media-{{media.length}}" id="media">
            {{#media}}
              <a target="_blank" href="{{imageUrl}}"><img src="{{imageUrl}}"></a>
            {{/media}}
          </section>
        {{/hasMedia}}

        {{#hasLinks}}
          <a href="{{link.url}}" target="_blank" rel="noopener" id="links">
            <img src="{{linkImageUrl}}">
            <section id="link-body">
              <h3>{{link.title}}</h3>
              {{#link.description}}<p>{{link.description}}</p>{{/link.description}}
              <span>{{linkHostname}}</span>
            </section>
          </a>
        {{/hasLinks}}

        <a href="{{twitterUrl}}" target="_blank" rel="noopener" id="created-at" class="tweet__link">
          <time datetime="{{createdAt.toISOString}}">{{createdAt.toLocaleString}}</time>
          via Twitter
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 400 400"><defs><style>.cls-1{fill:none;}.cls-2{fill:#1da1f2;}</style></defs><path class="cls-2" d="M153.62,301.59c94.34,0,145.94-78.16,145.94-145.94,0-2.22,0-4.43-.15-6.63A104.36,104.36,0,0,0,325,122.47a102.38,102.38,0,0,1-29.46,8.07,51.47,51.47,0,0,0,22.55-28.37,102.79,102.79,0,0,1-32.57,12.45,51.34,51.34,0,0,0-87.41,46.78A145.62,145.62,0,0,1,92.4,107.81a51.33,51.33,0,0,0,15.88,68.47A50.91,50.91,0,0,1,85,169.86c0,.21,0,.43,0,.65a51.31,51.31,0,0,0,41.15,50.28,51.21,51.21,0,0,1-23.16.88,51.35,51.35,0,0,0,47.92,35.62,102.92,102.92,0,0,1-63.7,22A104.41,104.41,0,0,1,75,278.55a145.21,145.21,0,0,0,78.62,23"/></svg>
        </a>

        <a href="https://www.heise.de/embetty?wt_mc=link.embetty.poweredby" target="_blank" rel="noopener" id="powered-by" title="embetty - displaying remote content without compromising your privacy.">
          powered by {{{embettyLogo}}}</a>
      </article>
    `
  }

  async becomesVisible() {
    await super.becomesVisible()

    const linkImage = this.shadowRoot.querySelector('#links img')
    if (linkImage) {
      linkImage.addEventListener('error', _e => {
        linkImage.remove()
      })
    }
  }
}

defineElement('embetty-tweet', Tweet)
