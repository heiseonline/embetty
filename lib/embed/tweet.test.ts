/* eslint-disable security/detect-non-literal-fs-filename */
import { createTweet, getFetchSpy } from '../../test/lib/util'
import Tweets from '@heise/embetty-base/test/tweets.json'
import './tweet'

beforeEach(() => {
  document.body.innerHTML = ''
  jest
    .spyOn(Date.prototype, 'toLocaleString')
    .mockReturnValue('10/03/2013, 15:07:53')
})

describe('Tweet', () => {
  it('should fetch remote data', async () => {
    const spy = await getFetchSpy(Tweets.text)
    const { element } = await createTweet(Tweets.text)
    expect(element._data.id_str).toEqual(Tweets.text)
    expect(element._fetched).toBeTruthy()
    expect(spy).toBeCalledWith('tweet/934029337019416579')
  })

  it('should generate the tweet header', async () => {
    await getFetchSpy(Tweets.text)
    const { query } = await createTweet(Tweets.text)
    expect(query('header img')).toBeTruthy()
    expect(query('header span strong')).toBeTruthy()
    expect(query('header span a')).toBeTruthy()
  })

  it('should generate the tweet body', async () => {
    await getFetchSpy(Tweets.text)
    const { query } = await createTweet(Tweets.text)
    expect(query('article')).toBeTruthy()
    expect(query('article p')).toBeTruthy()
    expect(query('article time')).toBeTruthy()
  })

  it('should show images', async () => {
    await getFetchSpy(Tweets.image)
    const { query } = await createTweet(Tweets.image)
    expect(query('#media img')).toBeTruthy()
  })

  it('should contain a time tag', async () => {
    await getFetchSpy(Tweets.text)
    const { query } = await createTweet(Tweets.text)
    const time = query('time')
    const datePattern = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d+Z$/
    expect(datePattern.test(time.getAttribute('datetime')!)).toBeTruthy()
    expect(time.textContent?.length).toBeGreaterThan(5)
  })

  it('a[target=_blank] should have a [rel=noopener] attribute', async () => {
    await getFetchSpy(Tweets.link)
    const { element } = await createTweet(Tweets.link)
    const links = [...element.shadowRoot!.querySelectorAll('a[target=_blank')]
    expect(links.length).toBe(4)

    links.forEach((link) => {
      expect(link.getAttribute('rel')).toBe('noopener')
    })
  })

  it('tweet link should contain the username and the tweet id', async () => {
    await getFetchSpy(Tweets.text)
    const { element } = await createTweet(Tweets.text)
    const links = [...element.shadowRoot!.querySelectorAll('.tweet__link')]
    expect(links).toHaveLength(1)
    expect(links[0].getAttribute('href')).toBe(
      'https://twitter.com/SiLVAFiSH/status/934029337019416579'
    )
  })

  describe('Tweet body', () => {
    it('should replace hash tags', async () => {
      await getFetchSpy(Tweets.text)
      const { query } = await createTweet(Tweets.text)
      const a = query(
        'article a[href="https://twitter.com/hashtag/Wochenende"]'
      )
      expect(a.textContent).toBe('#Wochenende')
    })

    it('should replace @user', async () => {
      await getFetchSpy(Tweets.answeredTweet)
      const { query, element } = await createTweet(Tweets.answeredTweet)
      const a = query('article a[href="https://twitter.com/VictorVEnciso"]')
      expect(element.shadowRoot?.innerHTML).toMatchSnapshot()
      expect(a.textContent).toBe('@VictorVEnciso')
    })

    it('should not output the image link when displaying the image', async () => {
      await getFetchSpy(Tweets.image)
      const { element } = await createTweet(Tweets.image)
      const text = element.fullText
      expect(text).not.toContain('https://t.co/wLy5Asq3z0')
    })

    it('should show tweet only without include-thread attribute', async () => {
      const { element } = await createTweet(Tweets.reply)
      expect(element.shadowRoot!.querySelector('embetty-tweet')).toBeNull()
    })

    it('should show thread and inherit include-thread attribute on being set', async () => {
      await getFetchSpy(Tweets.reply)
      const { element } = await createTweet(Tweets.reply, {
        'include-thread': '',
      })
      const parent = element.shadowRoot!.querySelector('embetty-tweet')
      expect(parent?.hasAttribute('include-thread')).toBeTruthy()
    })
  })
})
