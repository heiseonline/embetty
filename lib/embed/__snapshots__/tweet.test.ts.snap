// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tweet Tweet body should replace @user 1`] = `
"
    <style>tweet.scss mock</style>
      <header>
        <img src=\\"tweet/310768979788582912-profile-image\\">
        <span>
          <strong>Frantech Solutions</strong>
          <a href=\\"https://twitter.com/FrantechCA\\" target=\\"_blank\\" rel=\\"noopener\\">@FrantechCA</a>
        </span>
      </header>
      <article>
        <p><a href=\\"https://twitter.com/VictorVEnciso\\">@VictorVEnciso</a> I  got a ton of them coming up this coming week :) Check back then.</p>


        <a href=\\"https://twitter.com/FrantechCA/status/310768979788582912\\" target=\\"_blank\\" rel=\\"noopener\\" id=\\"created-at\\" class=\\"tweet__link\\">
          <time datetime=\\"2013-03-10T14:07:53.000Z\\">10/03/2013, 15:07:53</time>
          via Twitter
          <svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 400 400\\"><defs><style>.cls-1{fill:none;}.cls-2{fill:#1da1f2;}</style></defs><path class=\\"cls-2\\" d=\\"M153.62,301.59c94.34,0,145.94-78.16,145.94-145.94,0-2.22,0-4.43-.15-6.63A104.36,104.36,0,0,0,325,122.47a102.38,102.38,0,0,1-29.46,8.07,51.47,51.47,0,0,0,22.55-28.37,102.79,102.79,0,0,1-32.57,12.45,51.34,51.34,0,0,0-87.41,46.78A145.62,145.62,0,0,1,92.4,107.81a51.33,51.33,0,0,0,15.88,68.47A50.91,50.91,0,0,1,85,169.86c0,.21,0,.43,0,.65a51.31,51.31,0,0,0,41.15,50.28,51.21,51.21,0,0,1-23.16.88,51.35,51.35,0,0,0,47.92,35.62,102.92,102.92,0,0,1-63.7,22A104.41,104.41,0,0,1,75,278.55a145.21,145.21,0,0,0,78.62,23\\"></path></svg>
        </a>

        <a href=\\"https://www.heise.de/embetty?wt_mc=link.embetty.poweredby\\" target=\\"_blank\\" rel=\\"noopener\\" id=\\"powered-by\\" title=\\"embetty - displaying remote content without compromising your privacy.\\">
          powered by embetty.svg mock</a>
      </article>
    "
`;
